File: VB/E4755/Controllers/HomeController.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Web.Mvc
Imports DevExpress.Web.Mvc
Imports DevExpress.XtraReports.UI
Imports System.Drawing
Imports DevExpress.XtraPrinting.Shape

Namespace E4755.Controllers
	Public Class HomeController
		Inherits Controller


		Public Function Index() As ActionResult
			Return View()
		End Function

		Public Function Export() As ActionResult
			Dim model = Product.GetProducts()

			Dim gridViewState As MVCxGridViewState = CType(Session("gridViewState"), MVCxGridViewState)

			If gridViewState IsNot Nothing Then
				Dim generator As New MVCReportGeneratonHelper()
				AddHandler generator.CustomizeColumnsCollection, AddressOf generator_CustomizeColumnsCollection
				AddHandler generator.CustomizeColumn, AddressOf generator_CustomizeColumn
				Dim report As XtraReport = generator.GenerateMVCReport(gridViewState, model)
				generator.WritePdfToResponse(Response, "test.pdf", System.Net.Mime.DispositionTypeNames.Attachment.ToString())
				Return Nothing
			Else
				Return View("Index")
			End If
		End Function

		<ValidateInput(False)> _
		Public Function GridViewPartial() As ActionResult
			Dim model = Product.GetProducts()
			Return PartialView("_GridViewPartial", model)
		End Function

		Private Sub generator_CustomizeColumn(ByVal source As Object, ByVal e As ControlCustomizationEventArgs)
			If e.FieldName = "Discontinued" Then
				Dim control As New XRShape()
				control.SizeF = e.Owner.SizeF
				control.LocationF = New System.Drawing.PointF(0, 0)
				e.Owner.Controls.Add(control)
				control.Shape = New ShapeStar() With {.StarPointCount = 5, .Concavity = 30}
				AddHandler control.BeforePrint, AddressOf control_BeforePrint
				e.IsModified = True
			End If
		End Sub

		Private Sub control_BeforePrint(ByVal sender As Object, ByVal e As System.Drawing.Printing.PrintEventArgs)
			If Convert.ToBoolean((CType(sender, XRShape)).Report.GetCurrentColumnValue("Discontinued")) = True Then
				CType(sender, XRShape).FillColor = Color.Yellow
			Else
				CType(sender, XRShape).FillColor = Color.White
			End If
		End Sub

		Private Sub generator_CustomizeColumnsCollection(ByVal source As Object, ByVal e As ColumnsCreationEventArgs)
			e.ColumnsInfo(1).ColumnWidth *= 2
			e.ColumnsInfo(4).ColumnWidth += 30
			e.ColumnsInfo(3).ColumnWidth -= 30
			e.ColumnsInfo(e.ColumnsInfo.Count - 1).IsVisible = False
		End Sub

	End Class
End Namespace

File: CS/E4755/Controllers/HomeController.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Web.Mvc;
using DevExpress.Web.Mvc;
using DevExpress.XtraReports.UI;
using System.Drawing;
using DevExpress.XtraPrinting.Shape;

namespace E4755.Controllers {
    public class HomeController : Controller {
        

        public ActionResult Index() {
            return View();
        }

        public ActionResult Export() {
            var model = Product.GetProducts();

            MVCxGridViewState gridViewState = (MVCxGridViewState) Session["gridViewState"];

            if (gridViewState != null) {
                MVCReportGeneratonHelper generator = new MVCReportGeneratonHelper();
                generator.CustomizeColumnsCollection += new CustomizeColumnsCollectionEventHandler(generator_CustomizeColumnsCollection);
                generator.CustomizeColumn += new CustomizeColumnEventHandler(generator_CustomizeColumn);
                XtraReport report = generator.GenerateMVCReport(gridViewState, model);
                generator.WritePdfToResponse(Response, "test.pdf", System.Net.Mime.DispositionTypeNames.Attachment.ToString());
                return null;
            }
            else
                return View("Index");
        }

        [ValidateInput(false)]
        public ActionResult GridViewPartial() {
            var model = Product.GetProducts();
            return PartialView("_GridViewPartial", model);
        }

        void generator_CustomizeColumn(object source, ControlCustomizationEventArgs e) {
            if (e.FieldName == "Discontinued") {
                XRShape control = new XRShape();
                control.SizeF = e.Owner.SizeF;
                control.LocationF = new System.Drawing.PointF(0, 0);
                e.Owner.Controls.Add(control);
                control.Shape = new ShapeStar() {
                    StarPointCount = 5,
                    Concavity = 30
                };
                control.BeforePrint += new System.Drawing.Printing.PrintEventHandler(control_BeforePrint);
                e.IsModified = true;
            }
        }

        void control_BeforePrint(object sender, System.Drawing.Printing.PrintEventArgs e) {
            if (Convert.ToBoolean(((XRShape) sender).Report.GetCurrentColumnValue("Discontinued")) == true)
                ((XRShape) sender).FillColor = Color.Yellow;
            else
                ((XRShape) sender).FillColor = Color.White;
        }

        void generator_CustomizeColumnsCollection(object source, ColumnsCreationEventArgs e) {
            e.ColumnsInfo[1].ColumnWidth *= 2;
            e.ColumnsInfo[4].ColumnWidth += 30;
            e.ColumnsInfo[3].ColumnWidth -= 30;
            e.ColumnsInfo[e.ColumnsInfo.Count - 1].IsVisible = false;
        }
        
    }
}